File: ws01_plumbing/src/cpp01_topic/src/demo02_listener_stu.cpp

#include "rclcpp/rclcpp.hpp"
#include "base_interfaces_demo/msg/student.hpp"

// 创建发布方

using base_interfaces_demo::msg::Student;

class ListenerStu: public rclcpp::Node{
public:
    ListenerStu():Node("listener_stu"){
        subscription_ = this->create_subscription<Student>("chatter_stu", 10, std::bind(&ListenerStu::on_cb, this, std::placeholders::_1));
    }
private:
    void on_cb(const Student::SharedPtr stu)
    {
        RCLCPP_INFO(this->get_logger(), "订阅的消息: (%s,%d,%f)", stu->name.c_str(), stu->age, stu->height);
    }
    
    rclcpp::Subscription<Student>::SharedPtr subscription_;
};


int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ListenerStu>());
    rclcpp::shutdown();
    return 0;
}

File: ws01_plumbing/src/py01_topic/py01_topic/demo02_listener_stu.py


import rclpy
from rclpy.node import Node

from base_interfaces_demo.msg import Student

class ListenerStu(Node):
    def __init__(self):
        super().__init__("listener_stu_py")

        self.subscription = self.create_subscription(Student, "chatter_stu", self.on_cb, 10)

    def on_cb(self, msg):
        self.get_logger().warn("接收消息: %s %d %.2f" % (msg.name, msg.age, msg.height))

def main():
    rclpy.init()
    rclpy.spin(ListenerStu())
    rclpy.shutdown()
if __name__ == '__main__':
    main()